### PR TITLE
Simplify multitest storage

### PR DESCRIPTION
**Pull request by ethanfrey**
merge after #334 

Major overhaul of how storage and caching works in multi-test.
This should allow for much more extensibility in multi-test in the future

### CODE REVIEW

**Comment by ethanfrey** [comment]
Okay, this massive PR is ready for review.

It is a quasi-rewrite of the internals of multi-test, keeping a very similar API.

I think it makes everything much clearer and extensible, and lays the groundwork to easily add more "Keepers" to mock out staking, or handle custom messages

**Comment by maurolacy** [comment]
Can you add some context / description of the changes? It is impossible for me to review this in any meaningful way. Too many changes, that are unrelated, or at least, difficult to relate to the previous code base.

**Comment by ethanfrey** [comment]
> Can you add some context / description of the changes? It is impossible for me to review this in any meaningful way. Too many changes, that are unrelated, or at least, difficult to relate to the previous code base.

Yeah, I would just review the new codebase, and not the diff.
I will write a docs.rs file explaining the new architecture, then ask for a re-review.

The comments above were helpful to identify that.

**Comment by ethanfrey** [comment]
@maurolacy I added a DESIGN.md file that should help.

I would not review the diff, but rather review the new code by itself. You can add comments on the diff, but this is basically a re-write, so it makes more sense to look at it as a whole.

`test_helpers`, `transactions`, and `contracts` rust files are more or less unchanged.
`bank`, `wasm`, `app` and `executor` are > 50% modified

**Comment by maurolacy** [comment]
> @maurolacy I added a DESIGN.md file that should help.

Thanks. I'll take a look.

File: packages/multi-test/DESIGN.md [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+Note: This properly handles submessages and reply blocks.
+
+Note: While the API currently supports custom messages, we don't currently have a way to handle/process them.
+
+### Contracts
+
+Before you can call contracts, you must `instantiate` them. And to instantiate them, you need a `code_id`.
+In `wasmd`, this `code_id` points to some stored Wasm code that is then run. In multitest, we use it to
+point to a `Box<dyn Contract>` that should be run. That is, you need to implement the `Contract` trait
+and then add to to the app via `app.store_code(my_contract)`.
```
Comment (maurolacy) [inline packages/multi-test/DESIGN.md:57]:
```suggestion
and then add the contract to the app via `app.store_code(my_contract)`.
```

File: packages/multi-test/DESIGN.md [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+`sudo`, and `reply` (for submessages). Migration and IBC are currently not supported.
+
+In order to easily implement `Contract` from some existing contract code, we use the `ContractWrapper` struct,
+which takes some function pointers and combines them. You can look in `test_helpers.rs` for some examples
+or how to do so (and useful mocks for some test cases). Here is an example of wrapping a CosmWasm contract into
+a `Contract` trait to add to an `App`:
+
+```rust
+pub fn contract_reflect() -> Box<dyn Contract<CustomMsg>> {
+    let contract = ContractWrapper::new(execute_reflect, instantiate_reflect, query_reflect)
```
Comment (maurolacy) [inline packages/multi-test/DESIGN.md:69]: Might be good to mention that these fns are from the original contract code. Or show / list the imports above.

File: packages/multi-test/DESIGN.md [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+}
+```
+
+If you are not using custom messages in your contract, you can just use `dyn Contract<Empty>`.
+
+### Examples
+
+The best intro is most likely `integration.rs` in `cw20-escrow`, which shows sending and releasing native tokens in
+an escrow, as well as sending and releasing cw20 tokens. The first one updates the global bank ledger, the second
+actually shows how we can test orchestrating multiple contracts.
```
Comment (maurolacy) [inline packages/multi-test/DESIGN.md:81]: 👍🏼

File: packages/multi-test/DESIGN.md [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+    let api = Box::new(MockApi::default());
+    let bank = SimpleBank {};
+
+    App::new(api, env.block, bank, Box::new(MockStorage::new()))
+}
+```
+
+Note: This properly handles submessages and reply blocks.
+
+Note: While the API currently supports custom messages, we don't currently have a way to handle/process them.
```
Comment (maurolacy) [inline packages/multi-test/DESIGN.md:50]: Some early mention / reference to `Router` may be warranted here.

File: packages/multi-test/DESIGN.md [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+
+There is only one root Storage, stored inside `App`. This is wrapped into a transaction, and then passed down
+to other functions to work with. The code that modifies the Storage is divided into "Modules" much like the
+CosmosSDK. Here, we plan to divide logic into one "module" for every `CosmosMsg` variant. `Bank` handles `BankMsg`
+and `BankQuery`, `Wasm` handles `WasmMsg` and `WasmQuery`, etc.
+
+Each module produces a soon-to-be standardized interface to interact with. It exposes `execute` and `query` support
+as well as some "admin" methods that cannot be called by users but are needed for testcase setup. I am working on a
+design to make these "admin" methods more extensible as well. If you look at the two existing modules, you can
+see the great similarity in `query` and `execute`, such that we could consider making a `Module<MSG, QUERY>` trait.
```
Comment (maurolacy) [inline packages/multi-test/DESIGN.md:121]: 👍🏼

File: packages/multi-test/DESIGN.md [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+        amount: amount.to_vec(),
+      };
+      let res = router.execute(storage, block, sender.into(), msg.into())?;
+      Ok(res)
+    } else {
+      Ok(AppResponse::default())
+    }
+  }
+}
+```
```
Comment (maurolacy) [inline packages/multi-test/DESIGN.md:242]: Thanks. This explanation is really useful.

**Review by maurolacy — COMMENTED** [review]
Review of the DESIGN doc.

File: packages/multi-test/src/wasm.rs [hunk 7]
```diff
@@ -1,620 +1,706 @@
...
-    code_id: usize,
-    storage: &'a dyn Storage,
-}
-
-impl<'a> ContractRef<'a> {
-    fn new(code_id: usize, storage: &'a dyn Storage) -> Self {
-        ContractRef { code_id, storage }
-    }
-}
+pub trait Wasm<C>
```
[the remark this answers is not in the document]
Comment (maurolacy) [reply packages/multi-test/src/wasm.rs:40]: After the DESIGN doc, I think I get it. It takes a `Router` to be able to talk to other modules / contracts.

File: packages/multi-test/src/wasm.rs [hunk 8]
```diff
@@ -1,620 +1,706 @@
...
-    pub fn query(&self, querier: &dyn Querier, request: WasmQuery) -> Result<Binary, String> {
-        match request {
-            WasmQuery::Smart { contract_addr, msg } => {
-                self.query_smart(Addr::unchecked(contract_addr), querier, msg.into())
-            }
-            WasmQuery::Raw { contract_addr, key } => {
-                self.query_raw(Addr::unchecked(contract_addr), &key)
-            }
-            q => panic!("Unsupported wasm query: {:?}", q),
+impl<C> WasmKeeper<C>
```
Comment (maurolacy) [inline packages/multi-test/src/wasm.rs:150]: In passing: Shouldn't `SimpleBank` be callled `BankKeeper` for consistency?

**Review by maurolacy — APPROVED** [review]
lgtm, but this is still above my head. Need to get more exposure to both using and modifying / extending multi-test.

**Comment by ethanfrey** [comment]
Thanks for the feedback. I will update and merge.
There are a number of smaller steps in the next few weeks to get this closer to the vision

File: packages/multi-test/src/wasm.rs (hunk 8 above)
Comment (ethanfrey) [reply packages/multi-test/src/wasm.rs:150]: Done

**Pull request merged**